SpatialGDK/Source/SpatialGDKEditorCommandlet/Private/Commandlets/CookAndGenerateSchemaCommandlet.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+
+#include "CookAndGenerateSchemaCommandlet.h"
+#include "SpatialGDKEditorCommandletPrivate.h"
+#include "SpatialGDKEditorSchemaGenerator.h"
+
+DEFINE_LOG_CATEGORY(LogCookAndGenerateSchemaCommandlet);
+
+struct FObjectListener : public FUObjectArray::FUObjectCreateListener
+{
+	FObjectListener()
+	{
+		GUObjectArray.AddUObjectCreateListener(this);
+	}
+
+	~FObjectListener()
+	{
+	}
+public:
+	virtual void NotifyUObjectCreated(const class UObjectBase *Object, int32 Index) override
+	{
+		FSoftClassPath SoftClass = FSoftClassPath(Object->GetClass());
+		if (UnsupportedClasses.Contains(SoftClass))
+		{
+			return;
+		}
+		if (!VisitedClasses.Contains(SoftClass))
+		{
+			if (IsSupportedClass(Object->GetClass()))
+			{
+				UE_LOG(LogCookAndGenerateSchemaCommandlet, Verbose, TEXT("Object [%s] Created, Consider Class [%s] For Schema."),
+					*Object->GetFName().ToString(), *GetPathNameSafe(Object->GetClass()));
+				VisitedClasses.Add(SoftClass);
+			}
+			else
+			{
+				UnsupportedClasses.Add(SoftClass);
+			}
+		}
+	}
+
+
+	TSet<FSoftClassPath> VisitedClasses;
+	TSet<FSoftClassPath> UnsupportedClasses;
+};
+
+UCookAndGenerateSchemaCommandlet::UCookAndGenerateSchemaCommandlet()
+{
+	IsClient = false;
+	IsEditor = true;
+	IsServer = false;
+	LogToConsole = true;
+}
+
+UCookAndGenerateSchemaCommandlet::~UCookAndGenerateSchemaCommandlet()
+{
+	delete ObjectListener;
+	ObjectListener = nullptr;
+}
+
+int32 UCookAndGenerateSchemaCommandlet::Main(const FString& CmdLineParams)
+{
+	UE_LOG(LogCookAndGenerateSchemaCommandlet, Display, TEXT("Cook and Generate Schema Started."));
+	ObjectListener = new FObjectListener();
+	TSet<FSoftClassPath> ReferencedClasses;
+	
+	UE_LOG(LogCookAndGenerateSchemaCommandlet, Display, TEXT("Try Load Schema Database."));
+	if (!TryLoadExistingSchemaDatabase())
+	{
+		return 1;
+	}
+
+	UE_LOG(LogCookAndGenerateSchemaCommandlet, Display, TEXT("Finding supported C++ and in-memory Classes."));
+	for (TObjectIterator<UClass> ClassIt; ClassIt; ++ClassIt)
+	{
+		UClass* SupportedClass = *ClassIt;
+
+		if (IsSupportedClass(SupportedClass))
+		{
+			ReferencedClasses.Add(FSoftClassPath(SupportedClass));
+		}
+	}
+
+	UE_LOG(LogCookAndGenerateSchemaCommandlet, Display, TEXT("Starting Cook Command."));
+	int32 CookResult = Super::Main(CmdLineParams);
+	UE_LOG(LogCookAndGenerateSchemaCommandlet, Display, TEXT("Cook Command Completed."));
+
+	GUObjectArray.RemoveUObjectCreateListener(ObjectListener);
```

Comment:
Since we're adding Listener in ObjectListener constructor, I would prefer removing it in destructor then.